cmd/erasure-server-pool.go

```diff
@@ -47,6 +47,87 @@ import (
 	"github.com/minio/pkg/v3/wildcard"
 )
 
+// MultipartUpload represents multipart upload details to be cached
+type MultipartUpload struct {
+	Bucket   string
+	Object   string
+	UploadID string
+}
+
+// MultipartUploadCacheEntry represents one multipart upload cache entry
+type MultipartUploadCacheEntry struct {
+	Upload    MultipartUpload
+	Timestamp time.Time
+}
+
+// MultipartUploadCache represents a in memory cache to hold multipart upload details
+type MultipartUploadCache struct {
+	// Key would be uploadID as it's unique UUID
+	data  map[string]MultipartUploadCacheEntry
+	mutex sync.RWMutex
+}
+
+// NewMultipartUploadsCache initializes and returns a new Cache
+func NewMultipartUploadsCache(z *erasureServerPools) *MultipartUploadCache {
+	cache := &MultipartUploadCache{
+		data: make(map[string]MultipartUploadCacheEntry),
+	}
+	// Start the cleanup process
+	go cache.cleanExpiredEntries(z)
+	return cache
+}
+
+// Set adds a value to the cache with the current timestamp
+func (c *MultipartUploadCache) Set(key string, value MultipartUpload) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.data[key] = MultipartUploadCacheEntry{
+		Upload:    value,
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
+	expiry := globalAPIConfig.getStaleUploadsExpiry()
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	entry, exists := c.data[key]
+	if !exists || time.Since(entry.Timestamp) > expiry {
+		return MultipartUpload{}, false
+	}
+	return entry.Upload, true
+}
+
+// Has returns true if cache contains the key
+func (c *MultipartUploadCache) Has(key string) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	_, exists := c.data[key]
+	return exists
+}
+
+// Remove removes given entry from cache
+func (c *MultipartUploadCache) Remove(key string) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	delete(c.data, key)
+}
+
+// cleanExpiredEntries removes expired entries every hour
+func (c *MultipartUploadCache) cleanExpiredEntries(z *erasureServerPools) {
```

Comment:
Is it because, while Abort and Complete we are removing entries from cache?
Also not clear on usage of returned `[]string` of cleaned up upload ids from `cleanupStaleUploads`

Reply:
why not? if you have two then they will conflict, you only need one source of truth.